src/java.base/share/classes/java/lang/StringLatin1.java

```diff
@@ -824,6 +827,27 @@ static Stream<String> lines(byte[] value) {
         return StreamSupport.stream(LinesSpliterator.spliterator(value), false);
     }
 
+    static void putCharsAt(byte[] val, int index, int c1, int c2, int c3, int c4) {
+        assert index >= 0 && index + 3 < length(val) : "Trusted caller missed bounds check";
+        // Don't use the putChar method, Its instrinsic will cause C2 unable to combining values into larger stores.
```

Comment:
`StringLatin1` has no `putChar` method. I think you can just convert this to explicit array stores for the merge store optimization.

Reply:
explicit array stores  has the overhead of boundary checking. If putCharsAt of StringLatin1 is not implemented based on Unsafe, the performance will be worse than StringUTF16.

Of course, this is a common problem. StringUTF16.putChar is equivalent to Unsafe.putChar, without boundary checking. I found in many test scenarios that the UTF16 version performs better than the StringLatin1 version.

We may need to change some StringLatin1 related implementations to use Unsafe, otherwise users will turn off COMPACT_STRINGS to improve performance.

Reply:
Thank you for telling what the blocking issue is here. Does C2 not merge the bound checks when it does the merge stores? Interesting, and I think a fix from their side should be the way to go.

Reply:
MergeStore does not work when using StringUTF16.putChar, waiting for improvements from @eme64